zebra-scan/src/service/scan_task/scan.rs

```diff
@@ -125,31 +139,56 @@ pub async fn start(
             }
         }
 
-        let new_keys = ScanTask::process_messages(&cmd_receiver, &mut parsed_keys, network)?;
+        let (new_keys, new_result_senders) =
+            ScanTask::process_messages(&cmd_receiver, &mut parsed_keys, network)?;
+
+        // Send the latest version of `subscribed_keys` before spawning the scan range task
+        if !new_result_senders.is_empty() {
+            subscribed_keys.extend(new_result_senders);
+            // Ignore send errors, it's okay if there aren't any receivers.
+            let _ = subscribed_keys_sender.send(subscribed_keys.clone());
+        }
 
         // TODO: Check if the `start_height` is at or above the current height
         if !new_keys.is_empty() {
             let state = state.clone();
             let storage = storage.clone();
 
-            scan_task_sender
-                .send(ScanRangeTaskBuilder::new(height, new_keys, state, storage))
-                .await
-                .expect("scan_until_task channel should not be closed");
+            let start_height = new_keys
+                .iter()
+                .map(|(_, (_, _, height))| *height)
+                .min()
+                .unwrap_or(sapling_activation_height);
+
+            if parsed_keys.len() == new_keys.len() {
+                info!(?start_height, "setting new start height");
+                height = start_height;
```

Comment:
Why do we do this?

Reply:
> Why do we do this?

Earlier, it wasn't starting the scan task unless there were keys in the config, so there was always a start height. 

Now, it may start the scan task when there aren't any keys, and it would request blocks from the state service until it reaches the lowest start height of a parsed key.

This skips straight to the new min height without looping through all of the intermediate heights.

It also skips spawning a new task if the root scan task is free to scan the entire block height range for the new keys.

Reply:
I still don't see why we set the new `height` when the number of `parsed_keys` matches `new_keys`. What's different when the amounts differ? Also, could this not set `height` too high if a user sets `start_height` to a recent tip, for example?

Reply:
Ah, sorry. I should've seen it earlier. The original comparison makes sense to me now. Feel free to revert https://github.com/ZcashFoundation/zebra/commit/4d3099a9f4c9967fbfbf88bb1aaf6e8b8500355c.

Thank you for the answers.